# Repository Contents

This directory contains the following files and subdirectories:

## 

- **GeneInfo.txt**: Gene annotation information.
- **Eisenberg2013_Genes.txt**: Housekeeping gene from Eisenberg et al. 2013.
- **Lin2019_SingleCell.xlsx**: Gene stability index data from Lin et al. 2019.

## /Astrocytes

- **AstroVars.csv**: List of variables included in the astrocyte Random Forest models.
- **CRISPRi_Results Final.csv** and **CRISPRi_Power Simulation - Power per EGP.csv**: Results of the CRISPRi screening data generated in the present study.
  - *Scripts*: `CRISPRi_screen/2C_DE_EnhancerHits.csv` and `Power_Simulations.R` respectively.
  - Note: Positive EG pairs were defined as those significant (HitPermisive=TRUE) and downregulated (logFC.vst<0) from CRISPRi_Results Final.csv. Negative EG pairs were defined as those Well powered at a fold-change of 0.15 (WellPowered015=TRUE from CRISPRi_Power Simulation - Power per EGP.csv ) and not in the positive set. ** Supplementary Table 3F of the published article lists the Positive and Negative EG pair sets under the column "RandomForest" **.
- **TTseq_Results Table.csv**: TT-seq data generated in the present study.
  - *Script producing this file*: `Functional_annotation/EnhancerTranscription.R`
- **NHA_Peaks_Annotated.csv** and **NHA_All_ATACseq_Peaks_Annotated.csv**: ATAC-seq data generated in the present study for candidate enhancer peaks and all ATAC-seq peaks respectively.
  - *Script producing this file*: `CandidateEnhancerSelection_LibraryDesign/0a_PeakSelection.R`
- **TOBIAS_TF_counts.csv**: TF foot printing data generated in the present study.
  - *Script producing this file*: `Functional_annotation/Tobias_TFs/TobiasAnalysis.R`
- **MaxBelugaDiseaseScores.csv**: Disease impact scores for SNPs overlapping candidate enhancers, generated in the present study using the Beluga deep learning approach. Details in the methods section of Green et al. 2024.
  - *Script producing this file*: `Functional_annotation/Beluga/1_KnownVariants.R`
- **Astrocyte_Enh_HiC.csv**: HiC contact data from Hu et al Nat.Comm 2021 PMID 34172755.
  - *Script producing this file*: `./Scripts/InputDataProcessing/Enhancer_TSS_count_in_HiC.m`
- **ENCFF440FMQ.bed**: Bed file containing the overlap between candidate enhancers from the CRISPRi astrocyte screen and ENCODE rE2G predictions for astrocytes.
  - *Script producing this file*: `./Scripts/InputDataProcessing/OverlapENCODE_rE2G.sh`
- **RNAseq_FeatCounts.csv**: Gene expression data used for defining expressed genes when predicting significant EG pairs outside the candidate enhancer set. Count data is generated by ribodepleted RNAseq. TT-seq data is included as it was generated in the same sequencing run but not used for filtering.

### /Astrocytes/Processed_ChIP_ENCODE

ChIP-seq and DNase-seq data from NHA astrocytes, ChIP-seq data for K562 cells. BigWig files were downloaded from https://www.encodeproject.org/.
.tab files contain ChIP-seq/DNase-seq coverage of candidate enhancers from the CRISPRi astrocyte screen, generated using bigWigAverageOverBed (./Scripts/InputDataProcessing/Process_ChIP_DNase.sh).
The **Intergenic** subdirectory contains .tab files generated as above, for intergenic ATAC-seq peaks not tested in the CRISPRi screen.

- ENCFF058GEM: Chip.H3K27ac
- ENCFF476HBN: Chip.H3K27me3
- ENCFF242JDN: Chip.H3K4me1
- ENCFF184NZS: Chip.H3K4me3
- ENCFF702AYE: Chip.POLR2A
- ENCFF320CHE: DNase-seq

## /K562

- **K562Vars.csv**: List of variables included in astrocyte Random Forest models.
- **ENCODE-E2G_Predictions.tsv**: Training and prediction data for the ENCODE rE2G model downloaded from: https://github.com/karbalayghareh/ENCODE-E2G/tree/main/data/crispri
- **ENCODE-E2G_Extended_Predictions.tsv**: Training and prediction data for the ENCODE rE2G-extended model downloaded from https://github.com/karbalayghareh/ENCODE-E2G/tree/main/data/crispri
- **ENCFF968BZL.tsv**: ENCODE benchmarking dataset combining CRISPRi screening in K562 cells across several studies, downloaded from https://www.encodeproject.org/.

### /K562/Processed_ChIP_ENCODE

ChIP-seq and DNase-seq data from K562 cells. BigWig files were downloaded from https://www.encodeproject.org/.
.tab files contain ChIP-seq/DNase-seq coverage of candidate enhancers from the ENCODE K562 CRISPRi benchmarking dataset, generated using bigWigAverageOverBed (./Scripts/InputDataProcessing/Process_ChIP_DNase.sh).  

- ENCFF205FNC: DNase-seq
- ENCFF290LQY: Chip.H3K4me1
- ENCFF301TVL: Chip.H3K27ac
- ENCFF438GBD: Chip.POLR2A
- ENCFF611YPB: Chip.H3K4me3
- ENCFF915XIL: Chip.H3K27me3
